### DISCUSSION TITLE
[@types/react] 3-ary useReducer requires initialArg to fulfill State interface

### DISCUSSION THREAD

**Discussion by andoks** [DefinitelyTyped/DefinitelyTyped#68630] [Issues with a @types package] [answered]
When using the 3-ary useReducer call in React 17 (`useReducer<Reducer, State>( reducer, undefined, initializer)`), WebStorm claims that the initialArg (here undefined) must fulfill the State interface, while based on the [react docs](https://react.dev/reference/react/useReducer#reference) this is not necessary, as initialArg in such cases only is used as input to the initializer function to create the initial state.

A workaround is to cast undefined to any like so `useReducer<Reducer, State>( reducer, undefined as any, initializer)`, but this reduces type safety.

Is this a bug or am I misunderstanding something?

Dependencies and versions used that might be relevant
- @types/react-dom@17.0.25
- react@17.0.2
- react-dom@17.0.2
- react-scripts@5.0.1
- typescript@4.9.5

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment]
Can you provide a [TypeScript playground](typescript playground) highlighting the issue? WebStorm has a history of incorrect errors in their UI.

**Reply by andoks** [reply]
I will try

**Reply by andoks** [reply, marked as answer]
Indeed, when testing this in [codesandbox.io](https://codesandbox.io/p/sandbox/cranky-tu-mxhlf6?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clsu88hi100062866gctxlx23%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clsu88hi100022866w1wzwh8v%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clsu88hi100032866z21r7v03%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clsu88hlk000i28665c08hto8%2522%257D%252C%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clsu88hi100052866x3ytmcrr%2522%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clsu88hlk000i28665c08hto8%2522%253A%257B%2522id%2522%253A%2522clsu88hlk000i28665c08hto8%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522SANDBOX_ONBOARDING%2522%252C%2522id%2522%253A%2522sandbox-onboarding%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522id%2522%253A%2522clsu8pctn00022867np9pjxnt%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clsu8pctn00022867np9pjxnt%2522%257D%252C%2522clsu88hi100022866w1wzwh8v%2522%253A%257B%2522id%2522%253A%2522clsu88hi100022866w1wzwh8v%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clsu88hi100052866x3ytmcrr%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsu88hi10004286657ldnnm7%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clsu88hi100052866x3ytmcrr%2522%252C%2522activeTabId%2522%253A%2522clsu88hi10004286657ldnnm7%2522%257D%252C%2522clsu88hi100032866z21r7v03%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clsu88hi100032866z21r7v03%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) - it does not show up with the error. Will report issue to webstorm/jetbrains
